doc/man/testme.1
.TH TESTME "1" "January 2014" "testme" "User Commands"
.SH NAME
testme \- TestMe -- Utility to run unit tests
.SH SYNOPSIS
.B testme 
    \fB--chdir dir\fR
    \fB--clean\fR
    \fB--clobber\fR
    \fB--compile\fR
    \fB--continue\fR
    \fB--depth level\fR
    \fB--log logSpec\fR
    \fB--noserver\fR
    \fB--projects\fR
    \fB--rebuild\fR
    \fB--show\fR
    \fB--version\fR
    \fB--verbose\fR
    \fB--why\fR
    \fB[filters ...]\fR
.SH DESCRIPTION
TestMe is a dynamic unit test runner. It will traverse a directory tree and automatically build required unit tests and run them with output to the console.

.PP
Unit tests are stand-alone programs that have a '.tst' extension after the normal file extension. For example, a C unit test file would be named 'file.c.tst', while an Ejscript unit test file would be named 'file.es.tst' and a shell script test would be named 'file.sh.tst'.

.PP
Test parameters are passed to unit tests as environment strings with a 'TM_' prefix. 
Unit tests emit results to their standard output and are captured by TestMe. 
Unit tests emit results via a 'pass' or 'fail' command depending on the test results.
Unit tests can control the subsequent execution of tests via a 'skip' command and can write output to the TestMe console via
a 'write' command.
Unit tests can define values in the environment for subsequent tests by emitting a 'set' command (see below). 

.PP
TestMe will automatically compile and recompile C unit tests as required. Ejscript unit tests can be run directly without recompilation.

.SH OPERATION 
When TestMe is invoked, testme traverses the current directory tree looking for unit test files. 
At each directory level, it first runs any setup test files that have a '*.set' extension. The purpose of setup files
is to define the environment and configuration for all tests at or below that directory level. 
.PP
Next, TestMe will compile any Ejscript common files with a '*.es.com' extension. The purpose of common files it to define
shared code that will be used by Ejscript unit tests. Common files should wrap all script in a 'module' directive of the 
same name as the file without extensions. For example: 'support.es.com' would use 'module support { /* code */ }'. The 
compiled common file will be placed in a generated 'testme' directory.

.PP
For all C unit tests with a '*.c.tst' extension, TestMe will first create a 'testme' directory with a MakeMe file for
the unit test. It will then use MakeMe to build the unit test into a stand-alone executable.

.PP
When all is ready, TestMe will run all unit tests with a '*.tst' extension. For subdirectories, TestMe will recurse and
run unit tests in those subdirectories. The aggregated environment is passed down to unit tests in subdirectories. 
Note that if a unit tests uses the 'set' command to define a key value in the environment, it will only be passed to 
unit tests at that directory level or below.

.SH TEST ENVIRONMENT
TestMe communicates test parameters to unit tests via the environment.

.RS 5
 TM_NOSERVER -- Set if testme invoked with --noserver. Do not run server-side programs.
 TM_TOP      -- Set to the top directory for the project.
 TM_CFG      -- Set to the build output configuration directory.
 TM_DEPTH    -- Set to the test depth instructed via --depth.
 TM_PHASE    -- Set to 'Setup', 'Finalize' or 'Test'.
 TM_DIR      -- Set to the directory containing the test.
 TM_*        -- All defines in the me.h are converted to TM_ defines.
.RE
.PP

.SH TEST OUTPUT
A unit test should emit results to the standard out. The following commands are supported.

.RS 5
 fail reason ...
 info message ...
 pass 
 set key value
 skip reason ...
 write message ...
.RE
.PP
A 'fail' command will typically terminate a test run unless testme is invoked with --continue.
A 'pass' command will be counted and if no 'fail' commands are emitted by the unit test, the unit test will be PASSED.
An 'info' command will echo information message to the testme output. A 'write' message will write raw messages to the testme
output. A 'set' command will define a key in the environment that is passed to subsequent unit tests. The 'skip' command
will cause all subsequent unit tests in or below the current directory to be skipped.

.SH TESTME UNIT TEST C API
The follow C API is supported for C unit tests.
.RS 5

 bool  ttrue(expression);
 bool  tfalse(expression);
 bool  ttest(cchar *loc, cchar *expression, bool success);
 cchar *tget(cchar *key, cchar *def);
 int   tgetInt(cchar *key, int def);
 void  tinfo(cchar *fmt, ...);
 void  tset(cchar *key, cchar *value);
 void  tskip(bool skip);
 void  twrite(cchar *fmt, ...);
.RE
.PP

.SH OPTIONS
.TP
\fB\--chdir dir\fR
Change to the given directory before running tests.

.TP
\fB\--clean\fR
Clean the contents of all generated 'testme' directories.

.TP
\fB\--clobber\fR
Remove all generated 'testme' directories.

.TP
\fB\--compile\fR
Compile required C unit tests but do not run. Use --rebuild to force a recompile regardless of whether the unit test file has been updated or not.

.TP
\fB\--continue\fR
Continue to run tests despite any previous errors. Normal operation is to stop testing if any tests fail.

\fB\--depth level\fR
Set the unit test depth level.

.TP
\fB\--log logName[:logLevel]\fR
Specify a file to log test messages. TestMe will normally display test output to the console. The --log option will redirect this output to the specified log file. The log level
specifies the desired verbosity of output. Level 0 is the least verbose and level 5 is the most.

.TP
\fB\--noserver\fR
Do not run server side support code. This emits TM_NOSERVER into the environment for unit tests.

\fB\--projects\fR
Generate IDE projects for the specified unit tests. At least one test must be specified by name on the command line.
The IDE projects are generated in the 'testme' directory.

.TP
\fB\--rebuild\fR
Force a recompilation of all C unit tests.

.TP
\fB\--show\fR
Show the actual commands executed by TestMe.

.TP
\fB\--version\fR
Print the \fBejs\fR command version and exit.

.TP
\fB\--verbose\fR
Run in verbose mode with more trace about TestMe activities.

.TP
\fB\--why\fR
Display why various tests were run or not and why actions were taken.

.PP
.SH "REPORTING BUGS"
Report bugs to dev@embedthis.com.
.SH COPYRIGHT
Copyright \(co 2004-2014 Embedthis Software. TestMe and Ejscript are a trademarks of Embedthis Software.
.br
.SH "SEE ALSO"
me pak


src/tm/testme.c
/**
    testme.c - MakeMe unit test framework

    This program runs a script of the same name as this script.

    Copyright (c) All Rights Reserved. See details at the end of the file.
 */

/********************************** Includes **********************************/

#include    "ejs.h"

/*********************************** Locals ***********************************/

typedef struct App {
    Ejs     *ejs;
    cchar   *script;
} App;

static App *app;

static cchar *findTestMe();
static void manageApp(App *app, int flags);

/************************************ Code ************************************/

MAIN(ejsMain, int argc, char **argv, char **envp)
{
    Mpr         *mpr;
    Ejs         *ejs;
    EcCompiler  *ec;
    char        *argp, *searchPath, *homeDir, *logSpec, *traceSpec;
    int         nextArg, err, flags;

    /*  
        Initialize Multithreaded Portable Runtime (MPR)
     */
    mpr = mprCreate(argc, argv, 0);
    app = mprAllocObj(App, manageApp);
    mprAddRoot(app);
    mprAddStandardSignals();

    if (httpCreate(HTTP_CLIENT_SIDE | HTTP_SERVER_SIDE) < 0) {
        mprLog("me", 0, "Cannot create HTTP services");
        return EJS_ERR;
    }
    if (mprStart(mpr) < 0) {
        mprLog("me", 0, "Cannot start mpr services");
        return EJS_ERR;
    }
    argc = mpr->argc;
    argv = (char**) mpr->argv;
    err = 0;
    logSpec = 0;
    traceSpec = 0;
    searchPath = 0;

    for (nextArg = 1; nextArg < argc; nextArg++) {
        argp = argv[nextArg];
        if (*argp != '-') {
            break;
        }
        if (smatch(argp, "--chdir") || smatch(argp, "--home") || smatch(argp, "-C")) {
            if (nextArg >= argc) {
                err++;
            } else {
                homeDir = argv[++nextArg];
                if (chdir((char*) homeDir) < 0) {
                    mprLog("me", 0, "Cannot change directory to %s", homeDir);
                }
            }

        } else if (smatch(argp, "--debugger") || smatch(argp, "-D")) {
            mprSetDebugMode(1);

        } else if (smatch(argp, "--log")) {
            if (nextArg >= argc) {
                err++;
            } else {
                logSpec = argv[++nextArg];
            }

        } else if (smatch(argp, "--name")) {
            /* Just ignore. Used to tag commands with a unique command line */ 
            nextArg++;

        } else if (smatch(argp, "--search") || smatch(argp, "--searchpath")) {
            if (nextArg >= argc) {
                err++;
            } else {
                searchPath = argv[++nextArg];
            }

        } else if (smatch(argp, "--trace") || smatch(argp, "-t")) {
            if (nextArg >= argc) {
                err++;
            } else {
                traceSpec = argv[++nextArg];
            }

        } else if (smatch(argp, "--version") || smatch(argp, "-V")) {
            mprPrintf("%s\n", ME_VERSION);
            return 0;

        } else if (*argp == '-' && isdigit((uchar) argp[1])) {
            if (!logSpec) {
                logSpec = sfmt("stderr:%d", (int) stoi(&argp[1]));
            }
            if (!traceSpec) {
                traceSpec = sfmt("stderr:%d", (int) stoi(&argp[1]));
            }

        } else {
            /* Ignore */
        }
    }
    if (logSpec) {
        mprStartLogging(logSpec, MPR_LOG_CMDLINE);
    }
    if (traceSpec) {
        httpStartTracing(traceSpec);
    }
    if ((app->script = findTestMe()) == 0) {
        mprLog("me", 0, "Cannot find testme.es or testme.mod");
        return MPR_ERR_CANT_FIND;
    }
    argv[0] = (char*) app->script;
    if ((ejs = ejsCreateVM(argc, (cchar**) &argv[0], 0)) == 0) {
        return MPR_ERR_MEMORY;
    }
    mprStartDispatcher(ejs->dispatcher);
    app->ejs = ejs;
    if (ejsLoadModules(ejs, searchPath, NULL) < 0) {
        return MPR_ERR_CANT_READ;
    }
    flags = EC_FLAGS_BIND | EC_FLAGS_DEBUG | EC_FLAGS_NO_OUT | EC_FLAGS_THROW;
    if ((ec = ecCreateCompiler(ejs, flags)) == 0) {
        return MPR_ERR_MEMORY;
    }
    mprAddRoot(ec);
    ecSetOptimizeLevel(ec, 9);
    ecSetWarnLevel(ec, 1);
    if (ecCompile(ec, 1, (char**) &app->script) < 0) {
        if (flags & EC_FLAGS_THROW) {
            ejsThrowSyntaxError(ejs, "%s", ec->errorMsg ? ec->errorMsg : "Cannot parse script");
            ejsReportError(ejs, "Error in script");
        }
        err = MPR_ERR;
    } else {
        mprRemoveRoot(ec);
        if (ejsRunProgram(ejs, NULL, NULL) < 0) {
            ejsReportError(ejs, "Error in script");
            err = MPR_ERR;
        }
    }
    if (err) {
        mprSetExitStatus(err);
    }
    app->ejs = 0;
    ejsDestroyVM(ejs);
    mprDestroy();
    return mprGetExitStatus();
}


static void manageApp(App *app, int flags)
{
    if (flags & MPR_MANAGE_MARK) {
        mprMark(app->ejs);
        mprMark(app->script);
    }
}


static cchar *findTestMe()
{
    cchar    *path;

/* UNUSED
    if (mprPathExists("makeme", X_OK)) {
        path = "makeme/testme.mod"; 
        if (mprPathExists(path, R_OK)) {
            return sclone(path);
        }
        path = "makeme/testme.es"; 
        if (mprPathExists(path, R_OK)) {
            return sclone(path);
        }
    }
*/
    path = mprJoinPath(mprGetAppDir(), "testme.mod"); 
    if (mprPathExists(path, R_OK)) {
        return path;
    }
    path = mprJoinPath(mprGetAppDir(), "testme.es"); 
    if (mprPathExists(path, R_OK)) {
        return path;
    }
    return 0;
}

/*
    @copy   default

    Copyright (c) Embedthis Software LLC, 2003-2014. All Rights Reserved.

    This software is distributed under commercial and open source licenses.
    You may use the Embedthis Open Source license or you may acquire a 
    commercial license from Embedthis Software. You agree to be fully bound
    by the terms of either license. Consult the LICENSE.md distributed with
    this software for full details and other copyrights.

    Local variables:
    tab-width: 4
    c-basic-offset: 4
    End:
    vim: sw=4 ts=4 expandtab

    @end
 */
